src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -1718,45 +1756,61 @@ UpdateColumnToMatchingTargetEntry(Var *column, Node *flattenedExpr, List *target
 	foreach(targetEntryCell, targetEntryList)
 	{
 		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		NodeTag targetEntryType = nodeTag(targetEntry->expr);
+		bool doneProcessing = false;
 
-		if (IsA(targetEntry->expr, Var))
+		switch (targetEntryType)
 		{
-			Var *targetEntryVar = (Var *) targetEntry->expr;
-
-			if (IsA(flattenedExpr, Var) && equal(flattenedExpr, targetEntryVar))
+			case T_Var:
 			{
-				column->varno = 1;
-				column->varattno = targetEntry->resno;
+				Var *targetEntryVar = (Var *) targetEntry->expr;
+
+				if (IsA(flattenedExpr, Var) && equal(flattenedExpr, targetEntryVar))
+				{
+					column->varno = 1;
+					column->varattno = targetEntry->resno;
+					doneProcessing = true;
+				}
 				break;
 			}
-		}
-		else if (IsA(targetEntry->expr, CoalesceExpr))
-		{
-			/*
-			 * flatten_join_alias_vars() flattens full oter joins' columns that is
-			 * in the USING part into COALESCE(left_col, right_col)
-			 */
-			CoalesceExpr *targetCoalesceExpr = (CoalesceExpr *) targetEntry->expr;
 
-			if (IsA(flattenedExpr, CoalesceExpr) && equal(flattenedExpr,
-														  targetCoalesceExpr))
+			case T_CoalesceExpr:
+			case T_FuncExpr:
```

Comment:
Do we really need `T_FuncExpr `? I think we shouldn't need that. 

The reason you had to add seems like on `FlattenJoinVarsMutator ()` we return the node itself for two cases (e.g., `return (Node *) column` and `return node;`). Instead, we should return copies of them (e.g., `copyObject(node)`). Otherwise, we're losing the modifications on the node, which is `strip_implicit_coercions ()` in this case.

I think, if you do that, we should be able to undo the changes on `UpdateColumnToMatchingTargetEntry ` as well

Reply:
I tried to remove the need for FuncExpr.  For some reason `pull_var_clause_default` did not reduce `max(var)` into a `var` but kept it as function reference in one of the existing test cases. 

regarding copyObject, `expression_tree_mutator` already does copying itself before calling mutator function.  I will try copyObject and update it accordingly.

Reply:
query forcing me to have funcexpr in full_join test is
```sql
SELECT max(val1)
FROM test_table_1 LEFT JOIN test_table_3 USING(id, val1)


`val1` has conflicting declarations in joined tables. test_table_1 defines is as int, test_table_3 uses it as bigint.
GROUP BY test_table_1.id
ORDER BY 1;
```
However following is just fine
```sql
SELECT max(val1)
FROM test_table_1 FULL JOIN test_table_3 USING(id, val1)
GROUP BY test_table_1.id
ORDER BY 1;
```

Reply:
> regarding copyObject, expression_tree_mutator already does copying itself before calling mutator function. I will try copyObject and update it accordingly.

Oops, I realized something that actually could be dangerous/wrong with the expression mutation logic we've implemented. `expression_tree_mutator()` does not [mutate](https://github.com/postgres/postgres/blob/master/src/backend/nodes/nodeFuncs.c#L2479-L2488) `Var`s. That'd actually break our logic, no? Because you've switch from `while (Var belongs to join){..}`, now we'd exit on the first loop.

So, should we do something like the following?
```SQL
diff --git a/src/backend/distributed/planner/query_pushdown_planning.c b/src/backend/distributed/planner/query_pushdown_planning.c
index 3bba6f25..a12a8452 100644
--- a/src/backend/distributed/planner/query_pushdown_planning.c
+++ b/src/backend/distributed/planner/query_pushdown_planning.c
@@ -1656,8 +1656,8 @@ FlattenJoinVarsMutator(Node *node, FlattenJoinVarsMutatorContext *context)
                        Assert(column->varattno > 0);
                        newColumn = (Node *) list_nth(rte->joinaliasvars, column->varattno - 1);
                        Assert(newColumn != NULL);
-                       return expression_tree_mutator(newColumn, FlattenJoinVarsMutator,
-                                                                                  (void *) context);
+                       /* add a clarifying comment */
+                       return FlattenJoinVarsMutator(newColumn, context);
                }
                else
                {
@@ -1775,7 +1775,7 @@ UpdateColumnToMatchingTargetEntry(Var *column, Node *flattenedExpr, List *target
                        }
 
                        case T_CoalesceExpr:
-                       case T_FuncExpr:
+                       //case T_FuncExpr:
 
                                /*
                                 * FlattenJoinVars() flattens full oter joins' columns that is
```

(And, we'd add `copyObject` before returning the `Var`s because that might not come from expression_tree_mutator with the above logic.)